spopt/locate/frlm.py

```diff
@@ -0,0 +1,2089 @@
+        """
+        if not self.facility_combinations:
+            self.generate_facility_combinations()
+        
+        path_refueling_combinations = {}
+        
+        for od_pair, path in tqdm(self.flow_paths.items(), desc="Generating combinations"):
+            valid_combinations = []
+            
+            for combo in self.facility_combinations:
+                if self.check_path_refueling_feasibility(path, combo):
+                    valid_combinations.append(combo)
+            
+            path_refueling_combinations[od_pair] = valid_combinations
+
+            self.compute_refueling_frequency(od_pair[0], od_pair[1])
+        
+        self.path_refueling_combinations = path_refueling_combinations
+        
+        total_valid = sum(len(combinations) for combinations in path_refueling_combinations.values())
+        print(f"Generated {total_valid} valid combinations across {len(path_refueling_combinations)} OD pairs")
+        
+        return path_refueling_combinations
+    
+    def calculate_node_weights(self, method: str ='double') -> np.ndarray:
+        """
+        str, default='double'
+            Weight calculation method ('single' or 'double')
+        """
+        if not self.flows:
+            raise ValueError("Flows must be loaded first")
+
+        max_node = max(max(od[0], od[1]) for od in self.flows.keys())
+        node_weights = np.zeros(max_node + 1)
+        total_flow = sum(self.flows.values())
+        
+        if method == 'single':
+            # Origin-only weights (paper definition)
+            for (origin, destination), volume in self.flows.items():
+                node_weights[origin] += volume
+            node_weights = node_weights / total_flow
+            
+        elif method == 'double':
+            # MOSEL-style weights (both origin and destination)
+            for (origin, destination), volume in self.flows.items():
+                node_weights[origin] += volume
+                node_weights[destination] += volume
+            node_weights = node_weights / (total_flow * 2)
+            
+        else:
+            raise ValueError("Method must be 'single' or 'double'")
+        
+        self.node_weights = node_weights
+        print(f"Calculated node weights using {method} method")
+        return node_weights
+    
+    def solve(self, 
+         solver: str = "greedy", 
+         solver_instance: Optional[Any] = None,
+         max_time_seconds: Optional[int] = 600,
+         objective: str = "flow",
+         weight_method: str = "double",
+         seed: Optional[int] = None,
+         initialization_method: str = "empty", 
+         max_iterations: int = 100,
+         **kwargs) -> Dict:
+        """
+        Solve the FRLM problem with specified solver.
+        
+        Parameters:
+        -----------
+        solver : str, default="greedy"
+            Solver to use: "greedy", "pulp", "exact"
+        solver_instance : Optional[Any], default=None
```

Comment:
No. This should be just done using `solver`. if `solver == "greedy"`, we dispatch to `_solve_greedy()`. otherwise, we assume this is a `solver_instance`, and we send it to `self._solve_pulp()`. 

The type is `Union[str, pulp.Solver]`.